stan/math/rev/core/count_vars.hpp

```diff
@@ -0,0 +1,151 @@
+#ifndef STAN_MATH_REV_CORE_COUNT_VARS_HPP
+#define STAN_MATH_REV_CORE_COUNT_VARS_HPP
+
+#include <stan/math/prim/meta.hpp>
+#include <stan/math/rev/meta.hpp>
+#include <stan/math/rev/core/var.hpp>
+
+#include <utility>
+#include <vector>
+
+namespace stan {
+namespace math {
+
+template <typename VecVar, require_std_vector_vt<is_var, VecVar>* = nullptr,
+          typename... Pargs>
+inline size_t count_vars_impl(size_t count, VecVar&& x, Pargs&&... args);
```

Comment:
Sorry for bringing this up somewhat late.... but shouldn't we place the "*_impl" stuff into an "internal" namespace? This applies to the other tools as well which we are writing with an impl naming.

Reply:
Oh... count_vars is the only one with a "impl" thing - for good reasons as here you have to initialise to zero. So, I think moving the "*_impl" functions into an internal namespace, makes sense to me. Right?